Sources/NIOHTTP1/HTTPTypes.swift

```diff
@@ -176,33 +176,332 @@ struct HTTPHeader {
     let value: HTTPHeaderIndex
 }
 
+
+private extension UInt8 {
+    var isASCII: Bool {
+        return self <= 127
+    }
+
+    var maskedLowercase: UInt8 {
+        return self & 0xdf
+    }
+}
+
+
+/// Known HTTP header name. These should be created once and re-used as instances are memory intensive and "expensive" to create.
+public struct HTTPHeaderName {
+
+    /// Holds the bytes of the name + ': '
+    fileprivate let bytesWithSeperator: ContiguousArray<UInt8>
+    /// Holds the lowercased bases (masked) of the name.
+    fileprivate let lowerCaseBytes: ContiguousArray<UInt8>
+
+    /// Lowercase version of the name.
+    fileprivate let lowerCaseName: String
+
+    /// Original version of the name.
+    private let value: String
+
+    /// The number of bytes used by the name (without the seperator).
+    internal var count: Int {
+        return lowerCaseBytes.count
+    }
+
+    /// Create a new `HTTPHeaderName`.
+    ///
+    /// - Parameter value: The value of the header.
+    public init(_ value: String) {
+        precondition(!value.utf8.contains(where: { !$0.isASCII }), "name must be ASCII")
+        self.value = value
+        self.lowerCaseName = value.lowercased()
+        var array = ContiguousArray(value.utf8)
+        array.append(contentsOf: ": ".utf8)
+        self.bytesWithSeperator = array
+        self.lowerCaseBytes = ContiguousArray(value.utf8.map { $0.maskedLowercase })
+    }
+}
+
+extension HTTPHeaderName: Equatable {
+    public static func ==(lhs: HTTPHeaderName, rhs: HTTPHeaderName) -> Bool {
+        return lhs.lowerCaseName == rhs.lowerCaseName
+    }
+}
+
+extension HTTPHeaderName: CustomStringConvertible {
+    public var description: String {
+        return self.value
+    }
+}
+
+/// Defines commonly used HTTP header names.
+public extension HTTPHeaderName {
+    public static let aIM = HTTPHeaderName("A-IM")
```

Comment:
@weissi @Lukasa I wonder if we should use "lower-case" for all of the header names because of HTTP2 tho ... WDYT ?

Reply:
Yes, I think we should. TitleCase is silly.

Reply:
ok will do :)